text/0000-unified_coroutines.md

```diff
@@ -0,0 +1,502 @@
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Consider a function. In the olden days, also called a subroutine. This concept is at the core of every programming language, since it is extremely useful. The subroutine has a single entrypoint and a single exit point. Both of these points provide 
+an interface, which is used to transfer data. Upon entering, caller can pass data into the subroutine and upon exiting, the subroutine can return data back to the caller.
+
+Many programming languages(including Rust) also adopted more general concept, callled coroutine. The coroutine , or `Generator` in Rust terms differs from function/subroutine in a single way. It allows the coroutine to `suspend` itself, by storing its state, and `yield`ing control back to the caller, along with data. The caller can then repeatedly pass more data back into the coroutine, and `resume` it. This back-and-forth communication is an extremely useful tool for solving a wide class of problems.
+
+Except this is not the truth for Rust's coroutines. The `Generator`, as it  was introduced in order to provide a tool for implementing `async-await`, does not provide this functionality. In order to implement async-await feature, the generators were implemented in their most basic form. And in this form, Generators can't accept arguments.
+
+These issues severely lessen the usability of the generator feature, and are not difficult to solve.
+
+Current generators take the form of a closure, which contains at least a single yield statement.
+
+```rust
+let name = "World";
+let done = "Done";
+let gen = || {
+    yield "Hello";
+    yield name;
+    return done;
+};
+```
+And are used by calling the `resume` method on the generator.
+```rust
+println!("{:?}", gen.resume());
+println!("{:?}", gen.resume());
+println!("{:?}", gen.resume());
+```
+
+Which results in 
+```rust
+Yielded("Hello")
+Yielded("World")
+Finished("Done")
+```
+
+This RFC proposes the ability of a generator to take arguments with a syntax used by closures.
+```rust
+let gen = |name: &'static str| {
+    yield "Hello";                
+    yield name;
+    return name;
+}
+```
+
+Then, we propose a way to pass the arguments to the generator in the form of a tuple.
+```rust 
+println!("{:?}", gen.resume(("Not used")));
+println!("{:?}", gen.resume(("World")));
+println!("{:?}", gen.resume(("Done")));
+```
+Which would also result in:
+```rust
+Yielded("Hello")
+Yielded("World")
+Finished("Done")
+```
+Or expanded with values in between:
+```rust
+let gen = |name: &'static str| {
+    // name = "Not used"
+    yield "Hello";  // name is also dropped here    
+    // name = "World
```

Comment:
Dropped / not dropped is not precise terms because `name` is `Copy`. Maybe change it to a non-copy type?